File: artifacts/api-server/src/lib/rules-engine.ts
export type ShortfallRisk = {
  dominant_driver: string;
};

export type Recommendation = {
  action: string;
  detail: string;
  driver: string;
};

/**
 * Deterministic planning rules.
 *
 * This function deliberately owns the operational decision. Any future
 * natural-language layer may only phrase the returned recommendation; it must
 * not choose a different action or driver.
 */
export function recommend(shortfallRisk: ShortfallRisk): Recommendation {
  const driver = shortfallRisk.dominant_driver;
  const normalizedDriver = driver.trim().toLowerCase();

  if (normalizedDriver === "equipment downtime" || normalizedDriver === "downtime_hours") {
    return {
      action: "Redeploy standby equipment",
      detail: "Review the nearest underutilised fleet and stage one compatible unit before the next maintenance window.",
      driver: "Equipment downtime",
    };
  }

  if (normalizedDriver === "rainfall" || normalizedDriver === "rainfall_mm") {
    return {
      action: "Advance stockpiling",
      detail: "Build a controlled surface stockpile ahead of the monsoon-sensitive production window.",
      driver: "Rainfall",
    };
  }

  if (
    normalizedDriver === "reserve confidence" ||
    normalizedDriver === "local_reserve_confidence"
  ) {
    return {
      action: "Redirect development toward priority zone",
      detail: "Prioritise drilling and development toward the nearest high-confidence unexplored zone before committing new capacity.",
      driver: "Reserve confidence",
    };
  }

  return {
    action: "Continue standard monitoring",
    detail: "No single dominant driver is classified; continue standard monitoring while the next evidence update is prepared.",
    driver,
  };
}

File: module3_recommendations/rules_engine.py
"""Deterministic recommendation rules; no LLM decision-making is allowed."""


def recommend(shortfall_risk: dict) -> dict:
    driver = shortfall_risk["dominant_driver"]
    if driver in {"downtime_hours", "Equipment downtime"}:
        return {
            "action": "equipment_redeployment",
            "detail": "Redeploy available equipment from the nearest underutilised mine; review the maintenance schedule.",
            "driver": driver,
        }
    if driver in {"rainfall_mm", "Rainfall"}:
        return {
            "action": "advance_stockpiling",
            "detail": "Increase the pre-monsoon stockpile buffer for this mine given the upcoming rainfall forecast.",
            "driver": driver,
        }
    if driver in {"local_reserve_confidence", "Reserve confidence"}:
        return {
            "action": "redirect_development",
            "detail": "Current working-face confidence is declining; redirect development toward the nearest higher-confidence unexplored zone.",
            "driver": driver,
        }
    return {
        "action": "monitor",
        "detail": "No single dominant driver identified; continue standard monitoring.",
        "driver": driver,
    }